admin/services/data-access/group.data.service.js

```diff
@@ -65,8 +161,199 @@ groupDataService.update = async function (id, data) {
+    {
+      name: 'groupId',
+      value: groupId,
+      type: TYPES.Int
+    }
+  ]
+  const insertSql = []
+  for (let index = 0; index < pupilIds.length; index++) {
+    const pupilId = pupilIds[index]
+    insertSql.push(`(@pg${index},@pp${index})`)
+    params.push({
+      name: `pg${index}`,
+      value: groupId,
+      type: TYPES.Int
+    })
+    params.push({
+      name: `pp${index}`,
+      value: pupilId,
+      type: TYPES.Int
+    })
+  }
+
+  const sql = `
+  BEGIN TRY
+  BEGIN TRANSACTION GroupPupilUpdate
+    DELETE ${sqlService.adminSchema}.[pupilGroup] WHERE group_id=@groupId;
+    INSERT ${sqlService.adminSchema}.[pupilGroup] (group_id, pupil_id)
+    VALUES ${insertSql.join(',')}; 
+ COMMIT TRANSACTION GroupPupilUpdate
+END TRY
+
+BEGIN CATCH
+  IF (@@TRANCOUNT > 0)
+   BEGIN
+      ROLLBACK TRANSACTION GroupPupilUpdate
+      PRINT 'Error detected, all changes reversed'
+   END
+  DECLARE @ErrorMessage NVARCHAR(4000);
+    DECLARE @ErrorSeverity INT;
+    DECLARE @ErrorState INT;
+
+    SELECT @ErrorMessage = ERROR_MESSAGE(),
+           @ErrorSeverity = ERROR_SEVERITY(),
+           @ErrorState = ERROR_STATE();
+
+    -- Use RAISERROR inside the CATCH block to return
+    -- error information about the original error that
+    -- caused execution to jump to the CATCH block.
+    RAISERROR (@ErrorMessage, -- Message text.
+               @ErrorSeverity, -- Severity.
+               @ErrorState -- State.
+               );
+END CATCH
+  `
+  console.log(params)
+  console.log(sql)
+  return sqlService.modify(sql, params)
+}
+
+/**
+ * Get pupils filtered by schoolId and/or groupId.
+ * and grouped by group_id.
+ * @param schoolId
+ * @param groupId
+ * @returns {Promise<*>}
+ */
+groupDataService.sqlFindPupils = async (schoolId, groupId) => {
+  let params = [
+    {
+      name: 'schoolId',
+      value: schoolId,
+      type: TYPES.Int
+    }
+  ]
+
+  let sql = `SELECT p.[id], p.[foreName], p.[middleNames], p.[lastName], g.[group_id]
+    FROM ${sqlService.adminSchema}.[pupil] p 
+    LEFT JOIN ${sqlService.adminSchema}.[pupilGroup] g 
+      ON p.id = g.pupil_id 
+    WHERE p.school_id=@schoolId 
+    AND (g.group_id IS NULL`
+
+  if (groupId) {
+    params.push({
+      name: 'groupId',
+      value: groupId,
+      type: TYPES.Int
+    })
+    sql += ` OR g.group_id=@groupId`
+  }
+
+  sql += ')'
+
+  return sqlService.query(sql, params)
+}
+
+/**
+ * Soft-deletes a group.
+ * @deprecated use sqlMarkGroupAsDeleted
+ * @param id
+ * @returns {Promise<void>}
+ */
 groupDataService.delete = async function (id) {
   return Group.updateOne({'_id': id}, {$set: {'isDeleted': true}}).exec()
 }
 
+/**
+ * Soft deletes a group.
+ * @param {number} groupId the id of the group to mark as deleted
+ * @returns {Promise<*>}
+ */
+groupDataService.sqlMarkGroupAsDeleted = async (groupId) => {
+  const params = [
+    {
+      name: 'groupId',
+      value: groupId,
+      type: TYPES.Int
+    }
+  ]
+  const sql = `
+  BEGIN TRY
```

Comment:
If this transaction block is duplicated (as it seems to be) why not extract it into a wrapper and put it in `sqlService`?

Reply:
will add

Reply:
done